### ISSUE TITLE
namespace + protected definitions

### ISSUE THREAD

**Issue by fpvandoorn** [leanprover/lean3#426]
When we are in the namespace `foo`, a theorem `foo.bar` will overwrite theorem `X.bar` where `X` is any opened namespace. This is very useful. For example in namespace `int`, `add` refers to `int.add` and not to `nat.add`. 

However, if `foo.bar` is a _protected_ definition, `foo.bar` also overwrites `X.bar`. I think this shouldn't happen. With protected definitions we should always write out `foo.bar` completely, and never shorten it. 

For example, in `equiv.hlean` we define `is_equiv.transport` and `equiv.transport` as protected definitions, but then if we are in the namespace `equiv` or `is_equiv` we have to write `eq.transport` explicitly when referring to the "normal" transport.

**Comment by leodemoura** [comment]
This change would affect many files.
For example, we would be forced to write `nat.rec_on` instead of `rec_on` in the namespace `nat`.
Similar for all other recursors: `rec`, `cases_on`, `induction_on`, etc.
Are you ok with that?

I'm not against your suggestion. I think it makes sense. It also easier to explain the meaning of `protected` if we adopt your suggestion.

@avigad @javra: What do you think?

**Comment by fpvandoorn** [comment]
Ah, I didn't think about the fact that these automatically generated definitions are also protected. Still, I'm in favor of the change. Most of the time I already write `rec_on` with explicit namespace (often I'm in a different namespace anyway, for example recursing over `eq` anywhere in the HoTT library).

As alternative we could let the behavior depend on whether the short name already exists. For example if we are in namespace `nat` and `rec_on` doesn't refer to anything (which is likely, since `foo.rec_on` is typically protected), then it refers to `nat.rec_on`, even though that definition is protected. However, if we are in namespace `equiv` and `transport` already exists (as short form of `eq.transport`) then `transport` refers to `eq.transport`, not `equiv.transport`, since `equiv.transport` is protected and `eq.transport` isn't. This is really complicated to explain though, so I personally like my original proposal better.

**Comment by leodemoura** [comment]
> This is really complicated to explain though, so I personally like my original proposal better.

Me too.

**Comment by avigad** [comment]
I am always reluctant to ask Leo to fix things that aren't really broken,
but I do like the proposed change. It seems awkward that the behavior under
"namespace nat" is different from the behavior under "open nat". This seems
more uniform.

There is a nice side effect. We defined "theorem nat.is_decidable :
decidable nat" mainly because if we wrote "nat.decidable", there would be a
conflict with the type "decidable" inside the namespace nat. With the
proposed behavior, that conflict goes away.

On Tue, Feb 10, 2015 at 1:46 PM, Leonardo de Moura <notifications@github.com

**Comment by leodemoura** [comment]
> With the proposed behavior, that conflict goes away.

This is a nice side effect...

**Comment by leodemoura** [comment]
I've tried to port some library files. 
https://github.com/leodemoura/lean/commit/f5f76e301dea464cfbcb896910fcd40b0fb95e01
In some cases, it looks much more verbose...
What do you think?

**Comment by avigad** [comment]
Hmm... It looks o.k. to me, I think.

When we import a module, we can say "open nat (renaming nat.induction_on ->
induction_on)", right? Is there a way to define a temporary alias in a
namespace by hand, so that a user could say "alias induction_on
nat.induction_on"?

I don't feel very strongly about any of this.

On Tue, Feb 10, 2015 at 6:26 PM, Leonardo de Moura <notifications@github.com

**Comment by leodemoura** [comment]
> When we import a module, we can say "open nat (renaming nat.induction_on ->
> induction_on)", right?

Yes

> Is there a way to define a temporary alias in a
> namespace by hand, so that a user could say "alias induction_on
> nat.induction_on"?

We can write (it is quite verbose)

```
private definition induction_on [reducible] = @nat.induction_on
```

Another option is

```
local notation `ind_on` := @nat.induction_on
```

We also talked about an `abbreviation` command
https://github.com/leanprover/lean/issues/365

**Comment by avigad** [comment]
They all sound reasonable to me.

Does "local notation" have essentially the same effect as what happens when
we "open nat" and get to use "foo" instead of "nat.foo"? If so, I would
encourage that solution in the tutorial and documentation for someone who
gets tired of writing nat.induction_on in the namespace.

It still seems sensible to me that "open" and "namespace" have the same
behavior with respect to protected definitions.

On Tue, Feb 10, 2015 at 6:57 PM, Leonardo de Moura <notifications@github.com

**Comment by fpvandoorn** [comment]
The changes seem also good to me. It's a bit longer, but it is also a little clearer for the user what recursor is used.

@javra said he also is in favor of the change, and that he always writes the namespace explicitly when using `rec_on`.

**Comment by leodemoura** [comment]
When is a good time to make the changes?
The tutorial will break in many places.
@avigad Could you adjust the tutorial?

I can fix the libraries and tests

**Comment by avigad** [comment]
Yes, if you tell me when you make the changes, I will pull everything and
fix the tutorial.

I am also willing to help with the library and tests. E.g. for nat you
could define "abbreviation induction_on := ...", and I'll eliminate it
later.

On Wed, Feb 11, 2015 at 2:52 PM, Leonardo de Moura <notifications@github.com

**Comment by leodemoura** [comment]
> Yes, if you tell me when you make the changes, I will pull everything and
> fix the tutorial.

I pushed the new following new branch: 
https://github.com/leanprover/lean/tree/new_protected

You can get it as usual using `git fetch --all`
Then, you switch to this branch by executing

```
git checkout new_protected
```

The libraries have been fixed, but I still have to fix the tests.
I can do the tests. 
It would be great if you could fix the tutorial.

Thanks,
Leo

**Comment by avigad** [comment]
I am building it now. I'll make the changes in the current branch of the
tutorial, but I won't push the changes until the changes to the lean and
the library are in place.

On Wed, Feb 11, 2015 at 4:01 PM, Leonardo de Moura <notifications@github.com

**Comment by leodemoura** [comment]
Great!
You can make a pull request. I will merge them after I push the changes to lean/blessed.

**Comment by avigad** [comment]
I just wrote the explanation below in the inductive definitions chapter. I
realized that the idiom to "unprotect" a definition is to write e.g.

  open weekday (renaming cases_on -> cases_on)

That seems perfectly reasonable to me, but I thought I would mention it.

---

It is often useful to group definitions and theorems related to a
structure in a namespace with the same name. For example, we can put
the =number_of_day= function in the =weekday= namespace. We are then
allowed to use the shorter name when we open the namespace.

The names =rec_on=, =cases_on=, =induction_on=, and so on are
generated automatically, and they are /protected/ to avoid clashes; in
other words, those names are not shorted when the namespace
is open. You can explicitly declare the shorter identifiers as
abbreviations at any time, however. Or, you can "unprotect" them using
the =renaming= option when you open a namespace.

#+BEGIN_SRC lean
import data.nat
open nat

inductive weekday : Type :=
sunday : weekday,
monday : weekday,
tuesday : weekday,
wednesday : weekday,
thursday : weekday,
friday : weekday,
saturday : weekday

-- BEGIN
namespace weekday
  local abbreviation cases_on := @weekday.cases_on

  definition number_of_day (d : weekday) : nat :=
  cases_on d 1 2 3 4 5 6 7
end weekday

eval weekday.number_of_day weekday.sunday

section
  open weekday (renaming cases_on → cases_on)

  eval number_of_day sunday
  check cases_on
end
-- END
#+END_SRC

On Wed, Feb 11, 2015 at 4:15 PM, Leonardo de Moura <notifications@github.com

> wrote:
> 
> Great!
> You can make a pull request. I will merge them after I push the changes to
> lean/blessed.
> 
> From: Jeremy Avigad [mailto:notifications@github.com]
> Sent: Wednesday, February 11, 2015 1:14 PM
> To: leanprover/lean
> Cc: Leonardo de Moura
> Subject: Re: [lean] namespace + protected definitions (#426)
> 
> I am building it now. I'll make the changes in the current branch of the
> tutorial, but I won't push the changes until the changes to the lean and
> the library are in place.
> 
> On Wed, Feb 11, 2015 at 4:01 PM, Leonardo de Moura <
> notifications@github.com

**Comment by leodemoura** [comment]
> I realized that the idiom to "unprotect" a definition is to write e.g.
> 
> open weekday (renaming cases_on -> cases_on)

Yes.

We can also use

open weekday (cases_on)

However, we will be loading _only_ “cases_on”.

The renaming idiom loads _all_ unprotected, and unprotects “cases_on”

Leo

**Comment by avigad** [comment]
I just made the pull request. The fixes were easy. In many places, I had already written "bool.rec_on" and "list.rec_on" in the namespace. For some of the examples involving repeated induction on the natural numbers, I found it natural to say

  abbreviation induction_on := @nat.induction_on

which nicely illustrates the usage.

I still have other additions to make to the tutorial -- discussion of "protected" "abbreviation" etc. I will do that soon.
